### ISSUE TITLE
Memory leak: twisted.web.http.Request objects should call _cleanup() on lost connection

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11181]
|<img alt="jon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jon reported|
|-|-|
|Trac ID|trac#1621|
|Type|defect|
|Created|2006-04-03 02:59:16Z|

Presently, (as of [16539]) the code for the ```twisted.web.http.Request``` "```connectionLost```" method is:

```
    def connectionLost(self, reason):
        """connection was lost"""
        pass
```

If a connection is unexpectedly closed by the client, the ```Request``` object will not call its ```_cleanup``` method, introducing the potential for memory leaks.  This can be solved by changing the ```connectionLost``` method to:

```
    def connectionLost(self, reason):
        """connection was lost"""
        self._cleanup()
```

<details><summary>Searchable metadata</summary>

```
trac-id__1621 1621
type__defect defect
reporter__jon jon
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__MemoryLeak MemoryLeak
time__1144033156000000 1144033156000000
changetime__1232037784000000 1232037784000000
version__ 
owner__jon jon
cc__therve cc__exarkun
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="jon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jon commented|
|-|-|

This is also true of ```twisted.web.server.Request```, and that's actually a bit more insidious to deal with, since it needs to be handled in a ```Deferred errback``` call.

**Comment by twisted-trac** [comment]
|<img alt="jon's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|jon commented|
|-|-|

I just wanted to emphasize the severity of this problem since it hasn't received any attention in three months.

If a client terminates a connection to a ```Request``` that's serving a large amount of content (say 10MB), a client-side closure of the connection will result in a huge loss of memory.  If the client has downloaded 9MB of the 10MB content, all 9MB downloaded so far will be stored in the ```Request```'s sent content history (I apologize -- I don't have the name of the member that stores the content history on hand).  If the client closes the connection at that point, the ```Request``` will disappear into non-garbage-collected limbo and all 9MB of its sent content history will consequently be leaked.

The potential for large-scale memory leaking is huge with this issue, and it also poses a searious DOS threat.  A malicious user could repeatedly request resources and terminate the connection shortly before the content had been downloaded, rapidly starving the server of memory.

The fix here is _extremely_ simple and has the potential to prevent major problems.  I implore you to implement it at your earliest opportunity.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

I had a look at this, and I didn't manage to reproduce the problem. The fix in description can't work as-is, because server.Request doesn't call its parent in connectionLost. In every unittest I try, the request is correctly garbage collected.

Do you have a test, or a runnable example, that shows the problem ?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set owner to jon|
|-|-|

Can you provide detailed reproduction instructions for this issue?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [26019]) Close the request body file in Request.connectionLost; also drop the reference to the HTTPChannel; change server.Request.connectionLost so it calls up to http.Request.connectionLost; tests for all

refs [#3176](https://github.com/twisted/twisted/issues/3176)
refs [#11181](https://github.com/twisted/twisted/issues/11181)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set status to `closed`|
|-|-|

(In [26030]) Merge request-content-cleanup-3176

Author: exarkun
Reviewer: glyph
Fixes: [#11181](https://github.com/twisted/twisted/issues/11181)
Fixes: [#3176](https://github.com/twisted/twisted/issues/3176)

Explicitly close the file object containing the request body in the
request object's `connectionLost` method.  This frees the space
allocated for the request body as soon as the request object has
no further need for it, rather than waiting for the garbage
collector to get around to doing the same thing.

The reference the request holds to the HTTP protocol object is now
also broken in the `connectionLost` method to try to simplify the
garbage collector's job.
